### PR TITLE
Widget statistics logic

### PR DESCRIPTION
**Pull request by Yumfriez**

### CODE REVIEW

File: src/main/java/com/epam/ta/reportportal/core/widget/content/filter/ActivityFilterStrategy.java [hunk 1]
```diff
@@ -38,6 +44,19 @@
...
 	public Map<String, ?> buildFilterAndLoadContent(LoadContentStrategy loadContentStrategy, ReportPortalUser.ProjectDetails projectDetails,
 			Widget widget) {
-		return null;
+		UserFilter userFilter = widget.getFilter();
+		Filter filter = new Filter(userFilter.getTargetClass(), Sets.newHashSet(userFilter.getFilterCondition()));
+		filter = updateWithDefaultConditions(filter, projectDetails.getProjectId());
+		return loadContentStrategy.loadContent(widget.getContentFields(), filter, widget.getWidgetOptions(), widget.getItemsCount());
+	}
+
+	private Filter updateWithDefaultConditions(Filter filter, Long projectId) {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/filter/ActivityFilterStrategy.java]: as i see, this method is used in every filter strategy
will it be better to think about moving it in basic interface and call something like 'specifyDefaultConditions'

File: src/main/java/com/epam/ta/reportportal/core/widget/content/filter/ActivityFilterStrategy.java [hunk 2]
```diff
@@ -38,6 +44,19 @@
...
+		Set<FilterCondition> defaultConditions = Sets.newHashSet(new FilterCondition(Condition.EQUALS,
+				false,
+				String.valueOf(projectId),
+				"project_id"
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/filter/ActivityFilterStrategy.java]: use constants provided by Anton

File: src/main/java/com/epam/ta/reportportal/core/widget/content/filter/MostFailedFilterStrategy.java [hunk 3]
```diff
@@ -1,19 +1,40 @@
...
 import com.epam.ta.reportportal.entity.widget.Widget;
+import com.epam.ta.reportportal.ws.model.launch.Mode;
+import com.google.common.collect.Sets;
 import org.springframework.stereotype.Service;
 
 import java.util.Map;
+import java.util.Set;
 
 @Service
 public class MostFailedFilterStrategy implements BuildFilterStrategy {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/filter/MostFailedFilterStrategy.java]: this class is the same as GeneralFilterStrategy. Do we need it?

File: src/main/java/com/epam/ta/reportportal/core/widget/content/filter/CompareLaunchesFilterStrategy.java [hunk 4]
```diff
@@ -38,6 +46,19 @@
 	@Override
 	public Map<String, ?> buildFilterAndLoadContent(LoadContentStrategy loadContentStrategy, ReportPortalUser.ProjectDetails projectDetails,
 			Widget widget) {
-		return null;
+		UserFilter userFilter = widget.getFilter();
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/filter/CompareLaunchesFilterStrategy.java]: this class is the same as GeneralFilterStrategy. Do we need it?

File: src/main/java/com/epam/ta/reportportal/core/widget/content/filter/PassingRateFilterStrategy.java [hunk 5]
```diff
@@ -22,46 +22,44 @@
...
-
-		String launchName = launchNameOption.getValues()
-				.stream()
-				.findFirst()
-				.orElseThrow(() -> new ReportPortalException(ErrorType.UNABLE_LOAD_WIDGET_CONTENT,
-						"Incorrect widget option. Name shouldn't be empty"
-				));
-
-		Launch latestByName = launchRepository.findLatestByName(launchName);
+		UserFilter userFilter = widget.getFilter();
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/filter/PassingRateFilterStrategy.java]: this class is the same as GeneralFilterStrategy. Do we need it?

File: src/main/java/com/epam/ta/reportportal/core/widget/content/filter/UniqueBugFilterStrategy.java [hunk 6]
```diff
@@ -38,6 +44,19 @@
 	@Override
 	public Map<String, ?> buildFilterAndLoadContent(LoadContentStrategy loadContentStrategy, ReportPortalUser.ProjectDetails projectDetails,
 			Widget widget) {
-		return null;
+		UserFilter userFilter = widget.getFilter();
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/filter/UniqueBugFilterStrategy.java]: this class is the same as ActivityFilterStrategy, so could we combine them in one common strategy? Something like ProjectFilterStrategy

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/ActivityContentLoader.java [hunk 7]
```diff
@@ -22,22 +22,75 @@
...
+
+		Map<String, List<String>> options = WidgetContentUtils.GROUP_WIDGET_OPTIONS.apply(widgetOptions);
+		validateWidgetOptions(options);
+
+		String login = options.get(LOGIN).iterator().next();
+
+		User user = userRepository.findByLogin(login)
+				.orElseThrow(() -> new ReportPortalException(ErrorType.USER_NOT_FOUND, "User with login " + login + " was not found"));
+
+		List<String> activityTypes = Optional.ofNullable(options.get(ACTIVITY_TYPE))
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/ActivityContentLoader.java]: let's move this to validation method and add 'null or empty'

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/ActivityContentLoader.java [hunk 8]
```diff
@@ -22,22 +22,75 @@
...
+				.orElseThrow(() -> new ReportPortalException("Activity types must not be null"));
+
+		//		String conditionValue = activityTypes.stream().collect(Collectors.joining(", "));
+		//
+		//		FilterCondition filterCondition = new FilterCondition(Condition.IN, false, conditionValue, "value");
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/ActivityContentLoader.java]: legacy ?

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/BugTrendChartContentLoader.java [hunk 9]
```diff
@@ -22,22 +22,51 @@
...
+		List<LaunchStatisticsContent> result = widgetContentRepository.bugTrendStatistics(filter, fields, limit);
+		return singletonMap(RESULT, result);
+	}
+
+	/**
+	 * Validate provided content fields.
+	 *
+	 * @param contentFields Map of provided content.
+	 */
+	private void validateContentFields(Map<String, List<String>> contentFields) {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/BugTrendChartContentLoader.java]: let's add some more validation on required content fields keys. for example, if we know that content fields should  content either 'execution' or 'defects' key, we probably need to validate that.

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/CasesTrendContentLoader.java [hunk 10]
```diff
@@ -22,22 +22,52 @@
...
+
+		return singletonMap(RESULT, result);
 	}
+
+	/**
+	 * Validate provided content fields.
+	 *
+	 * @param contentFields Map of provided content.
+	 */
+	private void validateContentFields(Map<String, List<String>> contentFields) {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/CasesTrendContentLoader.java]: let's add some more validation on required content fields keys. for example, if we know that content fields should content either 'execution' or 'defects' key, we probably need to validate that.

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/ChartInvestigatedContentLoader.java [hunk 11]
```diff
@@ -22,22 +22,50 @@
...
 
+	@Autowired
+	private WidgetContentRepository widgetContentRepository;
+
 	@Override
-	public Map<String, ?> loadContent(List<String> contentFields, Filter filter, Set<WidgetOption> widgetOptions, int limit) {
-		return null;
+	public Map<String, ?> loadContent(Set<ContentField> contentFields, Filter filter, Set<WidgetOption> widgetOptions, int limit) {
+		Map<String, List<String>> fields = GROUP_CONTENT_FIELDS.apply(contentFields);
+		//		validateContentFields(fields);
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/ChartInvestigatedContentLoader.java]: what's wrong with this?

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/CumulativeChartContentLoader.java [hunk 12]
```diff
@@ -37,7 +37,7 @@
 public class CumulativeChartContentLoader implements LoadContentStrategy {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/CumulativeChartContentLoader.java]: remove this cause we are not going to implement it now

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchStatisticsChartContentLoader.java [hunk 13]
```diff
@@ -22,22 +22,52 @@
...
+
+		return singletonMap(RESULT, content);
+	}
+
+	/**
+	 * Validate provided content fields.
+	 *
+	 * @param contentFields Map of provided content.
+	 */
+	private void validateContentFields(Map<String, List<String>> contentFields) {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchStatisticsChartContentLoader.java]: let's add some more validation on required content fields keys. for example, if we know that content fields should content either 'execution' or 'defects' key, we probably need to validate that.

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesComparisonContentLoader.java [hunk 14]
```diff
@@ -22,22 +22,84 @@
...
+
+		String launchName = options.get(LAUNCH_NAME_FIELD).iterator().next();
+		Launch latestLaunch = launchRepository.findLatestByNameAndFilter(launchName, filter)
+				.orElseThrow(() -> new ReportPortalException(ErrorType.LAUNCH_NOT_FOUND, "No launch with name: " + launchName));
+
+		//TODO change the way of getting previous launch
+		Long latestLaunchId = latestLaunch.getId();
+
+		List<ComparisonStatisticsContent> result = widgetContentRepository.launchesComparisonStatistics(filter, fields, limit,
+				latestLaunchId - 1,
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesComparisonContentLoader.java]: wtf

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesComparisonContentLoader.java [hunk 15]
```diff
@@ -22,22 +22,84 @@
...
+				latestLaunchId
+		);
+		return singletonMap(RESULT, result);
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesComparisonContentLoader.java]: let's make this widget with possibility to compare as much launches as i want to.
not only two. so, i see this in the next way. client specifies launch name and limit. ws gives comparison of 'limited' launches by provided name back.

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesDurationContentLoader.java [hunk 16]
```diff
@@ -22,22 +22,51 @@
...
+		List<LaunchesDurationContent> result = widgetContentRepository.launchesDurationStatistics(filter, fields, limit);
+		return singletonMap(RESULT, result);
+	}
+
+	/**
+	 * Validate provided content fields.
+	 *
+	 * @param contentFields Map of provided content.
+	 */
+	private void validateContentFields(Map<String, List<String>> contentFields) {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesDurationContentLoader.java]: let's add some more validation on required content fields keys. for example, if we know that content fields should content either 'execution' or 'defects' key, we probably need to validate that.

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesTableContentLoader.java [hunk 17]
```diff
@@ -22,22 +22,51 @@
...
+		List<LaunchStatisticsContent> result = widgetContentRepository.launchesTableStatistics(filter, fields, limit);
+		return singletonMap(RESULT, result);
+	}
+
+	/**
+	 * Validate provided content fields.
+	 *
+	 * @param contentFields Map of provided content.
+	 */
+	private void validateContentFields(Map<String, List<String>> contentFields) {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesTableContentLoader.java]: let's add some more validation on required content fields keys. for example, if we know that content fields should content either 'execution' or 'defects' key, we probably need to validate that.

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LineChartContentLoader.java [hunk 18]
```diff
@@ -37,7 +37,7 @@
 public class LineChartContentLoader implements LoadContentStrategy {
 
 	@Override
-	public Map<String, ?> loadContent(List<String> contentFields, Filter filter, Set<WidgetOption> widgetOptions, int limit) {
+	public Map<String, ?> loadContent(Set<ContentField> contentFields, Filter filter, Set<WidgetOption> widgetOptions, int limit) {
 		return null;
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LineChartContentLoader.java]: what is the situation with this?

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/MostFailedContentLoader.java [hunk 19]
```diff
@@ -111,7 +117,8 @@ private void validateWidgetOptions(Map<String, List<String>> widgetOptions) {
 	 * @return Map of grouped content fields by first part. Expected only one value.
 	 */
 	private void validateContentFields(Map<String, List<String>> contentFields) {
-		BusinessRule.expect(contentFields, notNull()).verify(ErrorType.BAD_REQUEST_ERROR, "Content fields should not be null.");
+		BusinessRule.expect(MapUtils.isNotEmpty(contentFields), equalTo(true))
+				.verify(ErrorType.BAD_REQUEST_ERROR, "Content fields should not be empty");
 		BusinessRule.expect(contentFields.size(), Predicate.isEqual(1))
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/MostFailedContentLoader.java]: as we have change the structure of content fields, we should change this validation with something more suitable

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/NotPassedTestsContentLoader.java [hunk 20]
```diff
@@ -22,22 +22,51 @@
...
+		List<NotPassedCasesContent> result = widgetContentRepository.notPassedCasesStatistics(filter, fields, limit);
+		return singletonMap(RESULT, result);
+	}
+
+	/**
+	 * Validate provided content fields.
+	 *
+	 * @param contentFields Map of provided content.
+	 */
+	private void validateContentFields(Map<String, List<String>> contentFields) {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/NotPassedTestsContentLoader.java]: same

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/BugTrendChartContentLoader.java (hunk 9 above)
Comment (pbortnik) [reply src/main/java/com/epam/ta/reportportal/core/widget/content/loader/BugTrendChartContentLoader.java]: and provided more detailed docs on restrictions

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/CasesTrendContentLoader.java (hunk 10 above)
Comment (pbortnik) [reply src/main/java/com/epam/ta/reportportal/core/widget/content/loader/CasesTrendContentLoader.java]: and provided more detailed docs on restrictions

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchStatisticsChartContentLoader.java (hunk 13 above)
Comment (pbortnik) [reply src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchStatisticsChartContentLoader.java]: and provided more detailed docs on restrictions

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesDurationContentLoader.java (hunk 16 above)
Comment (pbortnik) [reply src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesDurationContentLoader.java]: and provided more detailed docs on restrictions

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesTableContentLoader.java (hunk 17 above)
Comment (pbortnik) [reply src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesTableContentLoader.java]: and provided more detailed docs on restrictions

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/MostFailedContentLoader.java (hunk 19 above)
Comment (pbortnik) [reply src/main/java/com/epam/ta/reportportal/core/widget/content/loader/MostFailedContentLoader.java]: and provided more detailed docs on restrictions

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/NotPassedTestsContentLoader.java (hunk 20 above)
Comment (pbortnik) [reply src/main/java/com/epam/ta/reportportal/core/widget/content/loader/NotPassedTestsContentLoader.java]: and provided more detailed docs on restrictions

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/PassedRatePerLaunchContentLoader.java [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+		BusinessRule.expect(widgetOptions.containsKey(LAUNCH_NAME_FIELD), Predicate.isEqual(true))
+				.verify(ErrorType.UNABLE_LOAD_WIDGET_CONTENT, LAUNCH_NAME_FIELD + " should be specified for widget.");
+	}
+
+	/**
+	 * Validate provided content fields.
+	 *
+	 * @param contentFields Map of provided content.
+	 */
+	private void validateContentFields(Map<String, List<String>> contentFields) {
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/PassedRatePerLaunchContentLoader.java]: and provided more detailed docs on restrictions

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/UniqueBugContentLoader.java [hunk 22]
```diff
@@ -23,21 +23,29 @@
...
 public class UniqueBugContentLoader implements LoadContentStrategy {
 
+	@Autowired
+	private WidgetContentRepository widgetRepository;
+
 	@Override
-	public Map<String, ?> loadContent(List<String> contentFields, Filter filter, Set<WidgetOption> widgetOptions, int limit) {
-		return null;
+	public Map<String, ?> loadContent(Set<ContentField> contentFields, Filter filter, Set<WidgetOption> widgetOptions, int limit) {
+		return singletonMap(RESULT, widgetRepository.uniqueBugStatistics(filter, GROUP_CONTENT_FIELDS.apply(contentFields), limit));
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/UniqueBugContentLoader.java]: validations?

File: src/main/java/com/epam/ta/reportportal/core/widget/content/WidgetContentUtils.java [hunk 23]
```diff
@@ -33,25 +34,19 @@
...
 	}
 
-	public static final Function<List<String>, Map<String, List<String>>> GROUP_CONTENT_FIELDS = contentFields -> Optional.ofNullable(
+	public static final Function<Set<ContentField>, Map<String, List<String>>> GROUP_CONTENT_FIELDS = contentFields -> Optional.ofNullable(
 			contentFields)
 			.orElseThrow(() -> new ReportPortalException(ErrorType.BAD_REQUEST_ERROR, "Content fields shouldn't be null"))
-			.stream()
-			.map(it -> it.split(SPLITTING_REGEX))
-			.collect(groupingBy(it -> it[0], mapping(it -> it[1].toUpperCase(), toList())));
+			.stream().collect(Collectors.toMap(ContentField::getFieldName, v -> Lists.newArrayList(v.getValues())));
```
Comment (avarabyeu) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/WidgetContentUtils.java]: could be replaced with groupingBy

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/ActivityContentLoader.java [hunk 24]
```diff
@@ -22,22 +22,85 @@
...
-		return null;
+	public Map<String, ?> loadContent(Set<ContentField> contentFields, Filter filter, Map<String, String> widgetOptions, int limit) {
+
+		validateWidgetOptions(widgetOptions);
+
+		Map<String, List<String>> fields = WidgetContentUtils.GROUP_CONTENT_FIELDS.apply(contentFields);
+		validateContentFields(fields);
+
+		String login = widgetOptions.get(LOGIN);
+		User user = userRepository.findByLogin(widgetOptions.get(LOGIN))
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/ActivityContentLoader.java]: put login var here

File: src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesComparisonContentLoader.java [hunk 25]
```diff
@@ -22,22 +22,98 @@
...
+	 * @param filter        Filter for launch search
+	 */
+	private void validateWidgetOptions(Map<String, String> widgetOptions, Filter filter) {
+		BusinessRule.expect(MapUtils.isNotEmpty(widgetOptions), equalTo(true))
+				.verify(ErrorType.BAD_REQUEST_ERROR, "Widget options should not be null.");
+
+		String launchName = widgetOptions.get(LAUNCH_NAME_FIELD);
+		BusinessRule.expect(launchName, StringUtils::isNotEmpty)
+				.verify(ErrorType.UNABLE_LOAD_WIDGET_CONTENT, LAUNCH_NAME_FIELD + " should be specified for widget.");
+		BusinessRule.expect(launchRepository.findLatestByNameAndFilter(launchName, filter), notNull())
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/widget/content/loader/LaunchesComparisonContentLoader.java]: don't need this

**Pull request merged**